presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -226,6 +219,43 @@ public synchronized void process(Iterable<QueryExecution> queries)
         }
     }
 
+    private boolean isLastKilledQueryGone()
+    {
+        boolean lastKilledQueryIsGone = (lastKilledQuery == null);
+
+        if (!lastKilledQueryIsGone) {
+            ClusterMemoryPool generalPool = pools.get(GENERAL_POOL);
+            if (generalPool != null) {
+                lastKilledQueryIsGone = generalPool.getQueryMemoryReservations().containsKey(lastKilledQuery);
+            }
+        }
+        return lastKilledQueryIsGone;
+    }
+
+    private void logQueryKill(QueryId killedQueryId, List<MemoryInfo> nodes)
+    {
+        if (!log.isDebugEnabled()) {
+            return;
+        }
+        log.debug("Query Kill Decision: Killed %s", killedQueryId);
+        for (MemoryInfo node : nodes) {
+            MemoryPoolInfo memoryPoolInfo = node.getPools().get(GENERAL_POOL);
+            if (memoryPoolInfo == null) {
+                continue;
+            }
+            StringBuilder nodeDescription = new StringBuilder();
+            nodeDescription.append("Query Kill Scenario: ");
+            nodeDescription.append("MaxBytes ").append(memoryPoolInfo.getMaxBytes());
+            nodeDescription.append("FreeBytes ").append(memoryPoolInfo.getFreeBytes() + memoryPoolInfo.getReservedRevocableBytes());
+            nodeDescription.append("Queries ");
+            for (Map.Entry<QueryId, Long> entry : memoryPoolInfo.getQueryMemoryReservations().entrySet()) {
+                nodeDescription.append(entry.getKey()).append('=').append(entry.getValue()).append(',');
+            }
+            nodeDescription.setLength(nodeDescription.length() - 1);
+            log.debug(nodeDescription.toString());
```

Comment:
I think we want to log a single entry with newlines, similar to how stack traces are logged. Logging multiple entries will cause different logging messages to get mixed up and make this harder to search for.